### PR TITLE
Fix: Avoid reloading extension

### LINKED ISSUE #4: Preference changes should be reflected immediately
Right now I just provide a Save button that reloads the extension with the new prefs.

But in [bug 1419940](https://bugzilla.mozilla.org/show_bug.cgi?id=1419940) (Firefox 59) I allowed `null` to remove window-specific data in case the user switches to "count tabs in all windows". With this I can avoid the reload, and then there is no problem if the changes are reflected immediately.

### LINKED ISSUE #17: Don't reload add-on when saving
I reload the add-on when saving in order to destroy previous data. However, this may affect the scroll position, which is annoying.

In [bug 1419940](https://bugzilla.mozilla.org/show_bug.cgi?id=1419940) (Firefox 59) I allowed `null` to remove data, I should use that instead if available.

This is necessary for #4.

### PR DESCRIPTION
**Pull request by Jackymancs4**
Hello @Loirooriol 
Among the many alternatives your extension is the one I prefer the most for the job of counting tabs, so thanks!

Here, this PR refactor the code in order to fix #17 
The code is quite simple, so I wont explain it here, but feel free to ask anything on review.

This solution also open an easy road to close #4 since you can add an `onChange` listener to every input on `options.html` and simply send a `reload` message to the background.

Tested on Firefox 68.0.1 and 69dev09

### CODE REVIEW

**Comment by Jackymancs4** [comment]
> Actually I already started a patch for this, but backwards compatibility was hard, e.g. before Firefox 59 it's not possible to use null in order to clear a previously set badge text.

I tested in Firefox 58 and changing 
```
browser.browserAction.setBadgeText({ text: null, windowId });
```
to 
```
browser.browserAction.setBadgeText({ text: "", windowId });
```
provided the intended behavior of the extension. I can change that to ensure compatibility.

> At this point I should probably raise the minimum version from 52 to 60 (I would like to include ESR 60, still supported by Mozilla).

I support this.

> Though going this way will make it more difficult to port the extension to Chromium. I don't know if I want to do that, but currently it should be kind of trivial.

I don't see why. Polyfilling browser will be mandatory in any case with the current code base, but except this there aren't many breaking differences that I know of. I can look into this if you want, but keeping it for the future seems a better approach.

The whole PR was made because I wasn't able apply two settings change in a row without manually reloading the options page. So this is what I'm aiming to fix.

File: background.js [hunk 1]
```diff
@@ -32,7 +32,7 @@ let numTabs = new Map();
 let lastTime = new Map();
 let allWindows = undefined;
 
-function svgDataIcon(text) {
+function svgDataIcon(text, prefs) {
```
[the remark this answers is not in the document]
Comment (Jackymancs4) [reply background.js]: Yes, that's a residue of me trying to minimize the reliance on global variables. There was no reason to change too much of the code base so I reverted almost everything. Will fix.

File: background.js [hunk 2]
```diff
@@ -49,20 +49,17 @@ function svgDataIcon(text) {
...
+  node.style.fontSize = `${14-l}px`;
+  if (l > 2) {
+    node.setAttribute("textLength", "100%");
+    node.setAttribute("lengthAdjust", "spacingAndGlyphs");
+  } else {
+    node.removeAttribute("textLength");
+    node.removeAttribute("lengthAdjust");
+  }
+  node.textContent = text;
+  return "data:image/svg+xml," + encodeURIComponent(serializer.serializeToString(doc));
```
[the remark this answers is not in the document]
Comment (Jackymancs4) [reply background.js]: Yes, while I personally think it's a micro optimization with overall little performance gain, I will implement a cache for this.

File: background.js [hunk 3]
```diff
@@ -90,9 +87,11 @@ function show(windowId, num = -1) {
   let title = prefs.titlePrefix + text;
   browser.browserAction.setTitle({ title, windowId });
   if (prefs.useBadge) {
+    browser.browserAction.setIcon({ path: { 16: "icon16.gif", 64: "icon.svg" }, windowId });
```
[the remark this answers is not in the document]
Comment (Jackymancs4) [reply background.js]: Do you mean calling:
```
browser.browserAction.setIcon({});
```
?

File: background.js [hunk 4]
```diff
@@ -90,9 +87,11 @@ function show(windowId, num = -1) {
...
     browser.browserAction.setBadgeText({ text, windowId });
   } else {
-    browser.browserAction.setIcon({ path: svgDataIcon(text), windowId });
+    browser.browserAction.setIcon({ path: svgDataIcon(text, prefs), windowId });
+    browser.browserAction.setBadgeText({ text: null, windowId });
```
[the remark this answers is not in the document]
Comment (Jackymancs4) [reply background.js]: Not sure what do you want to be cleared here.

File: background.js [hunk 5]
```diff
@@ -171,4 +173,22 @@ function increase(windowId, increment) {
...
     update(allWindows, tabs.length);
   }
-})();
+}
+
+browser.runtime.onMessage.addListener(function (request, sender, sendResponse) {
+  if (request === "reload") {
+    numTabs = new Map();
+    lastTime = new Map();
+    allWindows = undefined;
```
[the remark this answers is not in the document]
Comment (Jackymancs4) [reply background.js]: I see. Will fix.

File: background.js [hunk 6]
```diff
@@ -171,4 +173,22 @@ function increase(windowId, increment) {
     let tabs = await browser.tabs.query({});
     update(allWindows, tabs.length);
   }
-})();
+}
+
+browser.runtime.onMessage.addListener(function (request, sender, sendResponse) {
+  if (request === "reload") {
+    numTabs = new Map();
+    lastTime = new Map();
```
[the remark this answers is not in the document]
Comment (Jackymancs4) [reply background.js]: Will fix.

File: background.js [hunk 7]
```diff
@@ -171,4 +173,22 @@ function increase(windowId, increment) {
...
+    allWindows = undefined;
+
+    browser.tabs.onCreated.removeListener(tabOnCreatedListener);
+    browser.tabs.onRemoved.removeListener(tabOnRemovedListener);
+    browser.tabs.onAttached.removeListener(tabOnAttachedListener);
+    browser.tabs.onDetached.removeListener(tabOnDetachedListener);
+    browser.windows.onRemoved.removeListener(windowOnRemovedListener);
+
```
[the remark this answers is not in the document]
Comment (Jackymancs4) [reply background.js]: Can you explain how? I haven't thoroughly looked into it.

**Comment by Jackymancs4** [comment]
Hey @Loirooriol 
I just received the last update of the extension and that reminded me of this PR.
Are you interested in me fixing the conflicts and continuing the discussion about it?

**Comment by Loirooriol** [comment]
@Jackymancs4 Some time ago I started some patches to do something like this, let me check if/how they work before you start investing more time on this.

**Comment by Jackymancs4** [comment]
Ok, I'll wait for feedbacks.

**Comment by Loirooriol** [comment]
#17 should be fixed in version 3.1! Please check if it's working fine

**Pull request closed without merging**